### PR TITLE
Scoreboard's Sidebar better changing detection

### PR DESCRIPTION
**Pull request by JuggleStruggle**
This allows better detection as to when something inside the scoreboard gets updated so that it is shown. This also creates a sub-category in both the Hotbar and in Scoreboard to save space on the text and to categorize them in their own section. This might fix the hacky solution provided how DisguiseLib might end up crashing the mod as well, though this was not tested.

Renames:
* Scoreboard Objective -> Scoreboard Title

### CODE REVIEW

**Comment by Crendgrim** [comment]
Very well commented code, putting me to shame!
Some things I noticed on a cursory glance (will have a deeper look into it, and how it works, and whether it causes issues with DisguiseLib still soon): there's some formatting issues. Are you using Eclipse? IntelliJ defaulted to using spaces everywhere, while your new code uses tabs. That confuses github's code view. Also in the files you touched, the imports got reordered. Bit of an oddity. I wonder if there's some editorconfig file I can create to keep editors in line?

From a code perspective, it looks very nice, though. Thank you a lot for the contribution :)

**Comment by JuggleStruggle** [comment]
I do use Eclipse and used tabs rather than spaces and the reason why imports get reordered is because I used the same workspace as one of my public mods that I've released + my own style so maybe I could reorder the imports if you want them your way, though I might do an small fix to both things by targeting the changed classes.

**Review by Crendgrim — COMMENTED** [review]
Very nice! I played around with it a bit and couldn't find any problems.

Consider adding yourself to the authors list in `fabric.mod.json` if you'd like, as this is quite a substantial improvement to the mod!

File: src/main/java/mod/crend/autohud/component/ScoreboardComponentState.java [hunk 1]
```diff
@@ -1,93 +1,358 @@
...
+        // of "this" qualifier,
+        // so instead all of it had to be redone to allow it
+        super(component, () -> null, true);
+        super.newValueSupplier = this::createObjectiveAndUpdateCachedItems;
+        super.oldValue = super.newValueSupplier.get();
+        this.collectPlayerScores();
     }
 
-    public static ScoreboardObjective getObjective() {
+    @SuppressWarnings("resource")
```
[the remark this answers is not in the document]
Comment (JuggleStruggle) [reply src/main/java/mod/crend/autohud/component/ScoreboardComponentState.java:73]: Eclipse tells me that `MinecraftClient.getInstance()` is `Resource leak: '<unassigned Closeable value>' is not closed at this location` so I added that to suppress the warning, though I got it removed now.

File: src/main/java/mod/crend/autohud/component/ScoreboardComponentState.java [hunk 2]
```diff
@@ -1,93 +1,358 @@
...
+
+        if (cachedPlayerTeamAndScore == null) return;
+
+        if (!Objects.equals(cachedPlayerTeamAndScore.left(), team.getName())) {
+            final boolean cachedTeamIsEmpty = cachedPlayerTeamAndScore.left() == null;
+
+            // cachedTeam still can return null, so better be safe than sorry
+            Pair<Team, Integer> cachedTeam = cachedTeamIsEmpty ? 
+                    null : this.cachedTeams.get(cachedPlayerTeamAndScore.left());
+            boolean cachedTeamHasDifferences = cachedTeamIsEmpty || cachedTeam == null ?
```
[the remark this answers is not in the document]
Comment (JuggleStruggle) [reply src/main/java/mod/crend/autohud/component/ScoreboardComponentState.java:244]: Used your version and it turns out to be cleaner! :smile:

**Comment by Crendgrim** [comment]
Looks okay, I'd merge it if you think it's ready

**Comment by JuggleStruggle** [comment]
This should be the last set of changes so do try checking and see if everything's correct and if it is then go ahead and merge it! :smile:

File: src/main/java/mod/crend/autohud/component/ValueComponentState.java [hunk 3]
```diff
@@ -26,7 +27,7 @@ public void update() {
     }
 
     protected boolean doReveal(T newValue) {
-        return !component.config.active() || newValue != oldValue;
+        return !component.config.active() || !Objects.equals(newValue, oldValue);
```
Comment (JuggleStruggle) [inline src/main/java/mod/crend/autohud/component/ValueComponentState.java:30]: Fixed it for you :wink:

Comment (Crendgrim) [reply src/main/java/mod/crend/autohud/component/ValueComponentState.java:30]: Ah, I didn't see this. I chose the same solution. :)

File: src/main/java/mod/crend/autohud/component/ScoreboardComponentState.java [hunk 4]
```diff
@@ -45,13 +45,11 @@ public class ScoreboardComponentState extends ValueComponentState<ScoreboardObje
     public ScoreboardComponentState(Component component) {
         // the newValueSupplier in the use of super constructor does not allow the use
         // of "this" qualifier, so instead all of it had to be redone to allow it
-        super(component, () -> null, true);
-        super.newValueSupplier = this::createObjectiveAndUpdateCachedItems;
-        super.oldValue = super.newValueSupplier.get();
+        super(component, ScoreboardComponentState::createObjective, true);
```
Comment (Crendgrim) [inline src/main/java/mod/crend/autohud/component/ScoreboardComponentState.java:48]: Ah, much cleaner. :+1:

**Comment by SyberiaKey** [comment]
Can't wait for merging to translate this mod 😄

**Comment by Crendgrim** [comment]
Nice, thank you for dealing with this part! :)

**Pull request merged**